src/Raven.Server/Documents/Revisions/RevisionsStorage.cs

```diff
@@ -1495,14 +1496,50 @@ private async Task WriteRevertedRevisions(List<Document> list, OperationCancelTo
             list.Clear();
         }
 
-        private bool PrepareRevertedRevisions(DocumentsOperationContext writeCtx, Parameters parameters, RevertResult result, List<Document> list, OperationCancelToken token)
+        private bool PrepareRevertedRevisions(DocumentsOperationContext writeCtx, Parameters parameters, RevertResult result, List<Document> list, OperationCancelToken token, HashSet<string> collections = null)
         {
             using (_database.DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext readCtx))
             using (readCtx.OpenReadTransaction())
             {
-                var revisions = new Table(RevisionsSchema, readCtx.Transaction.InnerTransaction);
-                foreach (var tvr in revisions.SeekBackwardFrom(RevisionsSchema.FixedSizeIndexes[AllRevisionsEtagsSlice],
-                    parameters.LastScannedEtag))
+                IEnumerable<Table.TableValueHolder> tvrs = null;
+                if (collections != null)
+                {
+                    if (collections.Comparer != null && collections.Comparer != StringComparer.OrdinalIgnoreCase)
+                    {
+                        throw new InvalidOperationException("'collections' hashset must have an 'OrdinalIgnoreCase' comparer");
+                    }
+                    tvrs = Enumerable.Empty<Table.TableValueHolder>();
+                    foreach (var collection in collections)
+                    {
+                        if (collection == null)
+                        {
+                            var msg = "Tried to revert revisions in collection that is null";
+                            if (_logger.IsInfoEnabled)
+                                _logger.Info(msg);
+                            result.WarnAboutFailedCollection(msg);
+                            continue;
+                        }
+                        var collectionName = _documentsStorage.GetCollection(collection, throwIfDoesNotExist: false);
+                        if (collectionName == null)
+                        {
+                            var msg = $"Tried to revert revisions in the collection '{collection}' which does not exist";
+                            if (_logger.IsInfoEnabled)
+                                _logger.Info(msg);
+                            result.WarnAboutFailedCollection(msg);
+                            continue;
+                        }
+                        var tableName = collectionName.GetTableName(CollectionTableType.Revisions);
+                        var revisions = readCtx.Transaction.InnerTransaction.OpenTable(RevisionsSchema, tableName);
+                        tvrs = tvrs.Concat(revisions.SeekBackwardFrom(RevisionsSchema.FixedSizeIndexes[CollectionRevisionsEtagsSlice], parameters.LastScannedEtag));
```

Comment:
Since the `parameters.LastScannedEtag` is shard I wonder if this will work correctly with batches.

Collection `Bar` scanned until Etag 100, so collection `Bar` might skip things in Etag 90?

Do we have a test that involve several batches?

Reply:
you have 2 problems here when batches are involved.
1. The search for revision must be sorted otherwise you might skip some of them
2. Every new batch you will start from the very first collection which you might already completed.

Much healthier approach here would be not to mix collections, but revert one collection at a time